File: ios/StableChannels/StableChannels/Services/WebSocket/TransactionMatcher.swift
import Foundation

struct MatchResult: Equatable, Hashable {
    let target: String
    let isTxid: Bool
}

/// Pure matching logic — no actor isolation, no side effects.
/// All state is passed in, so the matcher is a deterministic function
/// of its inputs and trivially testable.
struct TransactionMatcher {
    func matchAll(
        trackedAddresses: Set<String>,
        trackedTxids: Set<String>,
        msg: MempoolWSMessage,
        tx: MempoolWSTransaction
    ) -> [MatchResult] {
        var results = [MatchResult]()
        var seen = Set<MatchResult>()

        func appendIfNew(_ res: MatchResult) {
            if seen.insert(res).inserted {
                results.append(res)
            }
        }

        // Direct address in response JSON
        if let respAddr = msg.address, trackedAddresses.contains(respAddr) {
            appendIfNew(MatchResult(target: respAddr, isTxid: false))
        }

        // Match output scriptpubkey_address
        if let vouts = tx.vout {
            for vout in vouts {
                if let addr = vout.scriptpubkeyAddress, trackedAddresses.contains(addr) {
                    appendIfNew(MatchResult(target: addr, isTxid: false))
                }
            }
        }

        // Match input txid (outspend of tracked funding txid)
        if let vins = tx.vin {
            for vin in vins {
                if let inputTxid = vin.txid, trackedTxids.contains(inputTxid) {
                    appendIfNew(MatchResult(target: inputTxid, isTxid: true))
                }
            }
        }

        // Match tracked txids directly
        if let respTxid = msg.txid, trackedTxids.contains(respTxid) {
            appendIfNew(MatchResult(target: respTxid, isTxid: true))
        }

        // Match bulk multi-address-transactions dictionary keys
        if let multi = msg.multiAddressTransactions {
            for (addr, txGroup) in multi {
                guard trackedAddresses.contains(addr) else { continue }
                if (txGroup.mempool?.contains(where: { $0.txid == tx.txid }) == true) ||
                    (txGroup.confirmed?.contains(where: { $0.txid == tx.txid }) == true) ||
                    (txGroup.removed?.contains(where: { $0.txid == tx.txid }) == true) {
                    appendIfNew(MatchResult(target: addr, isTxid: false))
                }
            }
        }

        return results
    }
}

File: android/app/src/main/java/com/stablechannels/app/services/websocket/TransactionMatcher.kt
package com.stablechannels.app.services.websocket

data class MatchResult(val target: String, val isTxid: Boolean)

class TransactionMatcher {
    fun matchAll(
        trackedAddresses: Set<String>,
        trackedTxids: Set<String>,
        msg: MempoolWSMessage,
        tx: MempoolWSTransaction,
    ): List<MatchResult> {
        val results = LinkedHashSet<MatchResult>()

        if (!msg.address.isNullOrBlank() && trackedAddresses.contains(msg.address)) {
            results.add(MatchResult(target = msg.address, isTxid = false))
        }

        tx.vout?.forEach { vout ->
            val addr = vout.scriptpubkeyAddress
            if (!addr.isNullOrBlank() && trackedAddresses.contains(addr)) {
                results.add(MatchResult(target = addr, isTxid = false))
            }
        }

        tx.vin?.forEach { vin ->
            val inputTxid = vin.txid
            if (!inputTxid.isNullOrBlank() && trackedTxids.contains(inputTxid)) {
                results.add(MatchResult(target = inputTxid, isTxid = true))
            }
        }

        if (!msg.txid.isNullOrBlank() && trackedTxids.contains(msg.txid)) {
            results.add(MatchResult(target = msg.txid, isTxid = true))
        }

        msg.multiAddressTransactions?.forEach { (addr, txGroup) ->
            if (!trackedAddresses.contains(addr)) {
                return@forEach
            }

            val inMempool = txGroup.mempool?.any { it.txid == tx.txid } == true
            val inConfirmed = txGroup.confirmed?.any { it.txid == tx.txid } == true
            val inRemoved = txGroup.removed?.any { it.txid == tx.txid } == true

            if (inMempool || inConfirmed || inRemoved) {
                results.add(MatchResult(target = addr, isTxid = false))
            }
        }

        return results.toList()
    }
}
